src/gadgets/curves/zeroable_affine/mod.rs

```diff
@@ -1,73 +1,277 @@
+    /// Initializes a new non-infinite affine point with the specified coordinates
+    fn new<CS>(cs: &mut CS, x: NF, y: NF) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        Self {
+            x,
+            y,
+            is_infinity: Boolean::allocated_constant(cs, false),
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the x-coordinate of the point
+    fn x(&self) -> &NF {
+        &self.x
+    }
+
+    /// Returns the y-coordinate of the point
+    fn y(&self) -> &NF {
+        &self.y
+    }
+
+    /// Initializes the point at infinity. x and y are set to zero, and is_infinity is set to true.
+    fn zero_point<CS>(cs: &mut CS, params: &Arc<NF::Params>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let zero_nf = NF::allocated_constant(cs, GC::Base::zero(), params);
 
         Self {
-            x: zero_nn.clone(),
-            y: zero_nn,
-            is_zero: boolean_true,
-            _marker: std::marker::PhantomData,
+            x: zero_nf.clone(),
+            y: zero_nf,
+            is_infinity: Boolean::allocated_constant(cs, true),
+            _marker: PhantomData,
         }
     }
 
-    pub fn same_x<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, other: &mut Self) -> Boolean<F> {
+    /// Adds two affine points on the curve.
+    fn add<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // If x's are not the same, adding via unequal_x method
+        let same_x = self.same_x(cs, other);
+        let boolean_false = Boolean::allocated_constant(cs, false);
+        if same_x == boolean_false {
+            return self.add_unequal_x(cs, other);
+        }
+
+        // If y's are the same, doubling the point
+        let same_y = self.same_y(cs, other).negated(cs);
+        if same_y == boolean_false {
+            return self.double(cs);
+        }
+
+        Self::zero_point(cs, self.x.get_params())
+    }
+
+    /// Subtracts two affine points on the curve.
+    fn sub<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut negated_other = other.negate(cs);
+        self.add(cs, &mut negated_other)
+    }
+
+    /// Multiplies the affine point by a scalar.
+    fn mul<CS>(&mut self, cs: &mut CS, scalar: &GC::Base) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut result = Self::zero_point(cs, self.x.get_params());
+        let mut temp = self.clone();
+
+        // Convert the scalar to bits
+        let scalar_bits = scalar
+            .into_repr()
+            .as_ref()
+            .iter()
+            .rev()
+            .flat_map(|byte| (0..8).rev().map(move |i| (byte >> i) & 1 == 1))
+            .collect::<Vec<_>>();
+
+        for bit in scalar_bits {
+            if bit {
+                result = result.add(cs, &mut temp);
+            }
+            temp.double(cs);
```

Comment:
this will result in another non-homogenous circuit structure - the addition should happen through conditional select

Reply:
Oh, I did not even mention that we have not removed the `zeroable_affine` implementation. Updated it with the `conditionally_select`, but I believe we will not use this part further.